File: back/app/models/building_component_monitoring.py
from django.db import models
from django.db.models.signals import pre_save
from django.dispatch import receiver

from app.base.models.base_models import ActiveManager, BaseEntityModelMixin
from app.models.comment import Comment
from app.util import format_decimal
from documents.base.base_models import BaseDocumentModel


class BuildingComponentMonitoring(BaseDocumentModel, BaseEntityModelMixin):
    class Meta(object):
        db_table = "building_component_monitoring"
        verbose_name = "Seguimiento de componente de construcción"
        verbose_name_plural = "Seguimientos de componente de construcción"

    objects = ActiveManager()

    execution_status = models.CharField("Estado de ejecución", max_length=20, null=True)
    quality_status = models.CharField("Estado cualitativo", max_length=20, null=True)

    expected_amount = models.DecimalField(
        "Monto previsto", max_digits=20, decimal_places=2, null=True
    )
    expected_end_date = models.DateField("Fecha de finalización prevista", null=True)

    paid_amount = models.DecimalField(
        "Monto real actual", max_digits=20, decimal_places=2, null=True
    )
    pending_amount = models.DecimalField(
        "Monto pendiente", max_digits=20, decimal_places=2, null=True
    )
    real_end_date = models.DateField("Fecha de finalización real", null=True)
    physical_progress_percentage = models.DecimalField(
        "Porcentaje de avance físico", max_digits=5, decimal_places=2, null=True
    )

    building_component = models.ForeignKey(
        "BuildingComponent",
        on_delete=models.CASCADE,
        related_name="building_component_monitorings",
    )
    project = models.ForeignKey(
        "Project", on_delete=models.CASCADE, related_name="project_building_monitorings"
    )
    project_order = models.IntegerField("Orden")
    comments = models.ManyToManyField(Comment)

    @property
    def total_amount(self):
        if not self.paid_amount and not self.pending_amount:
            return None
        return format_decimal((self.paid_amount or 0) + (self.pending_amount or 0))


@receiver(pre_save, sender=BuildingComponentMonitoring)
def bc_monitoring_pre_save(sender, instance, *args, **kwargs):
    if instance.execution_status == "finalizado":
        instance.pending_amount = 0


class BuildingComponentProgress(models.Model):
    class Meta:
        managed = False
        db_table = "bcm_progress"

    building_component_monitoring = models.OneToOneField(
        BuildingComponentMonitoring,
        on_delete=models.DO_NOTHING,
        primary_key=True,
        related_name="progress",
    )
    financial_progress_percentage = models.DecimalField(
        "Porcentaje de avance financiero", max_digits=5, decimal_places=2, null=True
    )
    financial_weight = models.DecimalField(
        "Peso financiero", max_digits=5, decimal_places=2, null=True
    )

    def __str__(self):
        return f"{self.__class__.__name__}: {self.id}"

File: front/src/buildingComponentMonitoring/model/BuildingComponentMonitoring.js
import {
    createBuildingComponent,
    building_component_api_adapter,
} from "buildingComponent/model";
import {comments_api_adapter, createComments} from "comment/model";
import {COMPONENT_EXECUTION_STATUS_COMPLETED} from "component/config";

class BuildingComponentMonitorings extends Array {}

const building_component_monitoring_api_adapter = buildingComponentMonitoring => {
    if (buildingComponentMonitoring["comments"]) {
        buildingComponentMonitoring["comments"] = createComments(
            comments_api_adapter(buildingComponentMonitoring["comments"])
        );
    }
    if (buildingComponentMonitoring["building_component"]) {
        buildingComponentMonitoring["building_component"] = createBuildingComponent(
            building_component_api_adapter({
                ...buildingComponentMonitoring["building_component"],
            })
        );
    }
    if (
        buildingComponentMonitoring["execution_status"] ===
        COMPONENT_EXECUTION_STATUS_COMPLETED
    ) {
        buildingComponentMonitoring["pending_amount"] = 0;
    }

    return buildingComponentMonitoring;
};

const building_component_monitoring_view_adapter = buildingComponentMonitoring => {
    delete buildingComponentMonitoring["folder"];
    delete buildingComponentMonitoring["featured_image"];
    delete buildingComponentMonitoring["created_by"];
    delete buildingComponentMonitoring["created_at"];
    delete buildingComponentMonitoring["updated_at"];
    delete buildingComponentMonitoring["updated_by"];

    delete buildingComponentMonitoring["comments"];

    return buildingComponentMonitoring;
};

const building_component_monitorings_api_adapter = buildingComponentMonitorings => {
    return buildingComponentMonitorings.map(building_component_monitoring_api_adapter);
};

const createBuildingComponentMonitorings = (data = []) => {
    const buildingComponentMonitorings = BuildingComponentMonitorings.from(
        data,
        buildingComponentMonitoring =>
            createBuildingComponentMonitoring(buildingComponentMonitoring)
    );
    return buildingComponentMonitorings;
};

const createBuildingComponentMonitoring = ({
    id = null,
    execution_status = "",
    execution_status_label = "",
    quality_status = "",
    quality_status_label = "",
    expected_amount = "",
    expected_end_date = "",
    paid_amount = "",
    pending_amount = "",
    total_amount = "",
    financial_progress_percentage = "",
    physical_progress_percentage = "",
    real_end_date = "",
    building_component = null,
    comments = [],
    project = null,
    folder = "",
    featured_image = "",
    created_by = "",
    created_at = null,
    updated_by = "",
    updated_at = null,
} = {}) => {
    const publicApi = {
        id,
        execution_status,
        execution_status_label,
        quality_status,
        quality_status_label,
        expected_amount,
        expected_end_date,
        paid_amount,
        pending_amount,
        total_amount,
        financial_progress_percentage,
        physical_progress_percentage,
        real_end_date,
        building_component,
        comments,
        project,
        folder,
        featured_image,
        created_by,
        created_at,
        updated_at,
        updated_by,
    };

    return Object.freeze(publicApi);
};

export {
    createBuildingComponentMonitoring as default,
    createBuildingComponentMonitorings,
    building_component_monitoring_api_adapter,
    building_component_monitorings_api_adapter,
    building_component_monitoring_view_adapter,
};
